### DISCUSSION TITLE
toolkit:CameraView  Will cause the program to automatically exit, iOS

### DISCUSSION THREAD

**Discussion by bbhxwl** [dotnet/maui#23983] [Q&A]
Will cause the program to automatically exit, iOS

<img width="1153" alt="image" src="https://github.com/user-attachments/assets/3e7a8cb1-ca39-4fb7-9e2e-6976096ebf46">

**Comment by stephenquan** [comment]
You have to supply more information on the repro steps and the iOS device you're testing on.

**Reply by bbhxwl** [reply]
[HomeBedCamera.zip](https://github.com/user-attachments/files/16489607/HomeBedCamera.zip)

**Reply by bbhxwl** [reply]
> You have to supply more information on the repro steps and the iOS device you're testing on.

Take 2 quick photos, and if you take 2 photos in 1 second, there will be an error.

**Comment by stephenquan** [comment]
I think both your description and your code highlights a reentrant problem of your code. A simple solution would be set `IsEnabled = false` whilst a capture is in progress to guard against a reentrant problem, e.g.

```xaml
<Grid VerticalOptions="End" RowDefinitions="30">
    <Button x:Name="CaptureButton" Grid.Row="1" Clicked="Button_OnClicked" Height="100" Width="100" Text="拍照"></Button>
</Grid>
```

```c#
    private async void CameraView_OnMediaCaptured(object? sender, MediaCapturedEventArgs e)
    {
        var uuid=await Utils.GetAverageHash(e.Media);
        Console.WriteLine(uuid);
        CaptureButton.IsEnabled = true;
    }

    private void Button_OnClicked(object? sender, EventArgs e)
    {
        CaptureButton.IsEnabled = false;
        Camera.CaptureImage(new CancellationTokenSource().Token);
    }
```

**Reply by bbhxwl** [reply]
> I think both your description and your code highlights a reentrant problem of your code. A simple solution would be set `IsEnabled = false` whilst a capture is in progress to guard against a reentrant problem, e.g.
> 
> ```
> <Grid VerticalOptions="End" RowDefinitions="30">
>     <Button x:Name="CaptureButton" Grid.Row="1" Clicked="Button_OnClicked" Height="100" Width="100" Text="拍照"></Button>
> </Grid>
> ```
> 
> ```cs
>     private async void CameraView_OnMediaCaptured(object? sender, MediaCapturedEventArgs e)
>     {
>         var uuid=await Utils.GetAverageHash(e.Media);
>         Console.WriteLine(uuid);
>         CaptureButton.IsEnabled = true;
>     }
> 
>     private void Button_OnClicked(object? sender, EventArgs e)
>     {
>         CaptureButton.IsEnabled = false;
>         Camera.CaptureImage(new CancellationTokenSource().Token);
>     }
> ```

Cannot avoid exceptions, can we only avoid them this way?

**Reply by stephenquan** [reply]
Of course, there is more than one way to solve this problem. Do you understand the problem is because `Camera.CaptureImage` is slower (takes more time) than the rate in which the user clicks `Button_OnClicked`?

So there are several solutions:
 1. Guard against reentrant code else you will have Camera.CaptureImage being initiated before the previous is completed
 2. Use the CancellationToken to abort previous run of Camera.CaptureImage before commence of a new one
 3. Increase the speed, or shorten the time of a Camera.CaptureImage (which can be achieved by lowering the capture resolution)

N.B. CameraView is part of the CommunityToolkit.Maui and this discussion should not be here, but over in https://github.com/CommunityToolkit/Maui/discussions
